keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,432 @@
+    effect: "NoSchedule"
+  nodeSelector:
+    beta.kubernetes.io/os: windows
+  containers:
+  - name: sqlpod
+    image: ddebroy/sqlsrv:latest
+    volumeMounts:
+    - name: csi-sql-vol
+      mountPath: C:\Data
+    env:
+    - name: ACCEPT_EULA
+      value: "y"
+    - name: sa_password
+      value: "*****"
+  volumes:
+    - name: csi-sql-vol
+      persistentVolumeClaim:
+        claimName: sqlpvc
+```
+
+### Implementation Details
+
+CSI Node Plugins listen on domain sockets and respond to CSI API requests sent over GRPC from a container orchestrator like Kubernetes. They are responsible all storage management operations scoped around a specific node that are typically necessary after a CSI Controller Plugin has finished provisioning a Persistent Volume and attached it to the node. In Kubernetes, the CSI Node API calls are invoked by the CSI In-tree Plugin in the kubelet as well as the CSI Node Driver Registrar. The CSI Node Driver Registrar interacts with the Kubelet Plugin Watcher and it is maintained by the Kubernetes CSI community as a side-car container for deployment in CSI Node Plugin pods.
+
+![Kubernetes CSI Components](https://raw.githubusercontent.com/kubernetes/community/master/contributors/design-proposals/storage/container-storage-interface_diagram1.png?raw=true "Kubernetes CSI Components")
+
+Support for Unix Domain Sockets has been introduced in Windows Server 2019 and works across containers as well as host and container as long as the processes running in containers are listening on the socket. If a process from within a container wishes to connect to a domain socket that a process on the host OS is listening on, Windows returns a permission error. This scenario however does not arise in the context of interactions between Kubelet, CSI Node Driver Registrar and CSI Node Plugin as these involve a process in a container listening on a domain socket (CSI Node Driver Registrar or CSI Node Plugin) that a process on the host (Kubelet) connects to.
+
+Golang supports domain socket operations for Windows since go version 1.12. It was found that in Windows, `os.ModeSocket` is not set on the `os.FileMode` associated with domain socket files in Windows. Therefore determining whether a file is a domain socket file using `os.ModeSocket` does not work on Windows.
+
+Based on the above, we can conclude that some of the fundamental support in the OS and compiler with regards to domain sockets in the context of CSI plugin discovery and a channel for API invocation is present in a stable state in Windows Server 2019 today. Although there are some observed limitations with respect to domain sockets in Windows Server 2019, they are not major blockers in the context of CSI Node Plugins. In the section below, we call out the components in the context of CSI Node Plugins in Kubernetes that will need to be enhanced to properly account for Windows paths and make use of domain sockets in Windows in a manner very similar to Linux.
+
```

Comment:
Did we consider instead of using the proxy to deploy the drivers uncontainerized completely? Might be worth noting somewhere if that was an option considered and why it's not the final design

Reply:
Will add a note here pointing to the alternative design sections later down.